### ISSUE TITLE
arch/x86/boot/compressed/Makefile hardcodes readelf

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/linux#771]
Reported-by: Dmitry Golovin <dima@golovin.in>

https://github.com/ClangBuiltLinux/continuous-integration/issues/73#issuecomment-552566001

this should use `$(READELF)` not `${CROSS_COMPILE}readelf`, which also needs to be defined in the top level Makefile and `export`ed.

**Comment by tpimh** [comment]
This is not 100% identical to #766 as there is no `READELF` variable in the top-level Makefile. Defining it shouldn't be a problem and I suggest changing all occurrences of `$(CROSS_COMPILE)readelf` to `$(READELF)`, so the patch would not be limited to x86_64.

**Comment by tpimh** [comment]
```patch
--- a/Makefile
+++ b/Makefile
@@ -414,6 +414,7 @@ STRIP		= $(CROSS_COMPILE)strip
 OBJCOPY		= $(CROSS_COMPILE)objcopy
 OBJDUMP		= $(CROSS_COMPILE)objdump
 OBJSIZE		= $(CROSS_COMPILE)size
+READELF		= $(CROSS_COMPILE)readelf
 PAHOLE		= pahole
 LEX		= flex
 YACC		= bison
@@ -472,7 +473,7 @@ GCC_PLUGINS_CFLAGS :=
 CLANG_FLAGS :=

 export ARCH SRCARCH CONFIG_SHELL BASH HOSTCC KBUILD_HOSTCFLAGS CROSS_COMPILE AS LD CC
-export CPP AR NM STRIP OBJCOPY OBJDUMP OBJSIZE PAHOLE LEX YACC AWK INSTALLKERNEL
+export CPP AR NM STRIP OBJCOPY OBJDUMP OBJSIZE READELF PAHOLE LEX YACC AWK INSTALLKERNEL
 export PERL PYTHON PYTHON2 PYTHON3 CHECK CHECKFLAGS MAKE UTS_MACHINE HOSTCXX
 export KBUILD_HOSTCXXFLAGS KBUILD_HOSTLDFLAGS KBUILD_HOSTLDLIBS LDFLAGS_MODULE

diff --git a/arch/x86/boot/compressed/Makefile b/arch/x86/boot/compressed/Makefile
index aa976adb7..f1e3f90a6 100644
--- a/arch/x86/boot/compressed/Makefile
+++ b/arch/x86/boot/compressed/Makefile
@@ -103,7 +103,7 @@ vmlinux-objs-$(CONFIG_EFI_MIXED) += $(obj)/efi_thunk_$(BITS).o
 quiet_cmd_check_data_rel = DATAREL $@
 define cmd_check_data_rel
 	for obj in $(filter %.o,$^); do \
-		${CROSS_COMPILE}readelf -S $$obj | grep -qF .rel.local && { \
+		${READELF} -S $$obj | grep -qF .rel.local && { \
 			echo "error: $$obj has data relocations!" >&2; \
 			exit 1; \
 		} || true; \
```
Formatting still to be fixed.

**Comment by nickdesaulniers** [comment]
Yep, that's what I'd expect.  What's the difference between `$(READELF)` vs `${READELF}`?

**Comment by tpimh** [comment]
I believe they are identical. I only used curly braces because `${CROSS_COMPILE}` in this line used it. Should it be changed to parentheses as this variable syntax seems to be more common in kernel Makefiles?

**Comment by nickdesaulniers** [comment]
https://stackoverflow.com/a/25185742/1027966

No difference, but the line you're modifying is the only line in the file that uses `${}`, the whole rest of the file uses `$()`, so this patch could be an opportunity to make the file internally consistent.

**Comment by nickdesaulniers** [comment]
submitted: https://lore.kernel.org/r/20191204225446.202981-1-dima@golovin.in/

**Comment by nickdesaulniers** [comment]
accepted: https://git.kernel.org/pub/scm/linux/kernel/git/masahiroy/linux-kbuild.git/commit/?h=fixes&id=eefb8c124fd969e9a174ff2bedff86aa305a7438

**Comment by nickdesaulniers** [comment]
mainline: https://github.com/torvalds/linux/commit/eefb8c124fd969e9a174ff2bedff86aa305a7438
